### DISCUSSION TITLE
Load VoiceData into Dexed VST3 plugin in DAW?

### DISCUSSION THREAD

**Discussion by probonopd** [probonopd/MiniDexed#589] [Q&A] [answered]
I am working on a lua [script](https://gist.github.com/probonopd/3e7e3cab7714663249dad4b2da3f0660) that might load multiple Dexed VST3 plugin instances into REAPER and configure them using MiniDexed performance.ini files.

__Update:__ It looks like Dexed currently ignores sysex (SYX) messages on the tracks it is playing.
I have opened https://github.com/asb2m10/dexed/issues/401 for this.

![image](https://github.com/probonopd/MiniDexed/assets/2480569/f5d20c5e-6250-4ffd-8e69-b2de45f9c136)

**Comment by BobanSpasic** [comment]
I've just tested - Dexed can't read VCED dumps, just VMEM.

To load the voice data into Dexed you need to do the following:
- make an empty voice bank (4104 bytes) with proper header
- take the 8 voice data strings and convert from VCED to VMEM format
- put the new 8 VMEMs into the empty bank
- calculate and write the bank checksum (but I think Dexed does not check the checksum at all)
- load this bank into Dexed

**Reply by probonopd** [reply]
Are you saying that the Dexed VST3 plugin cannot receive individual voices, only whole banks of voices via MIDI sysex messages? I had tried to ask the same thing in https://github.com/asb2m10/dexed/discussions/352 and looking at this code

https://github.com/asb2m10/dexed/blob/bda3dc2aebfe528d43e66ac7bfef8f828daca9f4/Source/PluginProcessor.cpp#L618-L626

it seems like it should be able to load a individual voice if the sysex message is 156 bytes long?

**Reply by BobanSpasic** [reply]
I think that receiving single voices works. You can try with Synthmata.
Loading single voices from HDD does not work - that's what I tested an hour ago.

**Reply by probonopd** [reply]
Thanks a lot for your help.

So just to be sure, `43 0B 0B 2A 63 58 00 00 02 00 3A 03 01 06 00 00 49 00 05 00 0C 45 0A 0D 2A 63 58 00 00 22 08 09 03 00 01 00 04 4C 00 01 00 09 4E 49 33 37 63 3A 00 00 33 0C 0B 03 03 01 00 02 56 01 06 09 06 40 15 09 2C 63 4F 00 00 1E 00 00 00 00 07 00 03 63 00 01 00 08 42 0B 00 27 63 4F 00 00 20 00 09 00 00 03 00 02 4A 00 01 01 02 40 15 09 2C 63 4F 00 00 1E 00 00 00 00 07 00 02 63 00 01 00 04 63 63 63 63 32 32 32 32 08 07 00 23 00 00 00 01 00 02 18 6D 44 78 50 69 61 6E 6F 4C 52 3F` which we have in MiniDexed performance.ini is a VCED? What about leading `F0` and trailling `F7` which REAPER automatically inserts when one wants to send the above as a sysex on a MIDI track?

You still think I need to convert the VCED to VMEM in order to be able to send it as a sysex on a MIDI track? Or does that restriction only apply to loading `.syx` files?

**Reply by BobanSpasic** [reply]
Wait, you need a proper header:
`F0 43 00 00 01 1B`
and two bytes after the VCED: the checksum, and F7
As I think that Dexed does not do anything with checksum, just put anything there, just keep it smaller than F0 (max EF)

**Reply by BobanSpasic** [reply]
There is one thing that bother me - the single voice SysEx is 155 + 8 (header and tail). The byte 156 is never transmitted as a part of the SysEx for real Yamahas.
VolcaFM does transmit it instead of the checksum (here the incompatibility between VolcaFM and DX)

**Reply by probonopd** [reply]
So would you happen to know how I need to change/format the above voice to Dexed VST3 in REAPER as a sysex on a MIDI track?
* Prepend `F0 43 00 00 01 1B` to the long string above
* Append `00 F7` (the `00` being the checksum which doesn't matter)? Or just append `F7` because the `3F` in the string above is in the byte into which the checksum would normally go? (this way we get 163 bytes but it doesn't seem to work)

**Reply by BobanSpasic** [reply]
Byte 3 in the header should match the MIDI channel (00 to 0F) your Dexed is on.
It should work with just appending F7.

**Comment by probonopd** [comment]
Can't seem to get it to work... here is my Reaper file:

[LoadingVoiceDataDoesNotWork.zip](https://github.com/probonopd/MiniDexed/files/13627469/LoadingVoiceDataDoesNotWork.zip)

To see the sysex, double click the item e.g., on the first track, then click the third icon from the left in the toolbar to see the MIDI messages. Double click the sysex one to see its contents.

Playing this file will play the notes, but will not cause the voice contained in the sysex to be loaded into Dexed.

Do you have an idea why?

**Reply by probonopd** [reply]
More investigation over at https://github.com/asb2m10/dexed/discussions/352.

**Comment by probonopd** [comment, marked as answer]
Update: It looks like Dexed currently ignores sysex (SYX) messages on the tracks it is playing.
I have opened https://github.com/asb2m10/dexed/issues/401 for this.

**Reply by BobanSpasic** [reply]
I understand, it is a problem with VST version of Dexed. As I am only using the stand-alone Dexed, I could not see this problem.

**Reply by probonopd** [reply]
Thanks for your help anyway @BobanSpasic.

**Comment by maxcomperatore** [comment]
Two key technical hurdles explain why this is behaving like that between Reaper, Dexed, and sysex:

1. The VST3 specification deliberately dropped arbitrary SysEx streaming:
Steinberg explicitly removed raw MIDI 1.0 byte streaming (0xF0 ... 0xF7) from the base VST3 spec. VST3 replaced it with normalized parameter changes. While some DAWs (like Reaper) can package sysex into VST3 event queues, most VST3 plugins simply drop sysex on track playback unless specifically coded with a custom IDataEvent queue. If you switch Reaper to load the VST2 build of Dexed instead, the raw MIDI stream bypasses that constraint and passes sysex directly to the plugin instance.

2. VCED vs VMEM format differences:
A single voice edit buffer dump (VCED) is 155 bytes unpacked (one byte per parameter, envelopes, algorithm, feedback, LFO, and the 10-char ASCII name). 
A 32-voice bank (VMEM) is 4,096 bytes of packed voice data (32 voices x 128 bytes each) plus the 6-byte Yamaha header `F0 43 00 09 20 00`, 1-byte checksum, and `F7` (total 4,104 bytes).
In VMEM, pairs of parameters (like coarse frequency and osc mode, or detune and fine frequency) are bit-packed into single bytes. If Dexed receives an unpacked 155-byte VCED stream when expecting a packed 4,104-byte bank, it discards it without an error message.

We dealt with this exact conversion headache when building the Web MIDI engine for [bipluk](https://bipluk.com). If you ever need to inspect whether a DX7 sysex bank has valid checksums, properly packed VMEM blocks, or clean ASCII names, you can drop the .syx file right into the browser to verify the structure.
